### PR TITLE
Expose the resolution of capabilities for general use in protocol negotiations

### PR DESCRIPTION
**Pull request by buddhisthead**
The general mechanism of capabilities' preferred mode resolution can be exported to enable non-gossip usage,
such as over socket connections, e.g. protocols built on top of the connection and service managers (also in core). The first step is to refactor the resolver of preferred modes and export API such that the gen_server's state is not entangled in the core functions.

A separate PR will use the exported API to negotiate capabilities over a socket based protocol.

Tagged @jtuple for possible review as well.

### CODE REVIEW

File: src/riak_core_capability.erl [hunk 1]
```diff
@@ -179,6 +182,12 @@ update_ring(Ring) ->
             add_supported_to_ring(node(), Supported, Ring)
     end.
 
+%% @doc Make a capbility from
```
Comment (cmeiklejohn) [inline src/riak_core_capability.erl]: Missing end of sentence?

File: src/riak_core_capability.erl [hunk 2]
```diff
@@ -393,6 +402,23 @@ add_supported_to_ring(Node, Supported, Ring) ->
             {true, Ring2}
     end.
 
+%% Given my node's capabilities, my node's registered default modes, the
```
Comment (cmeiklejohn) [inline src/riak_core_capability.erl:408]: Missing `@doc`.

File: src/riak_core_capability.erl (hunk 1 above)
Comment (buddhisthead) [inline src/riak_core_capability.erl]: Addressed by c4bfd01

File: src/riak_core_capability.erl (hunk 2 above)
Comment (buddhisthead) [inline src/riak_core_capability.erl:408]: Addressed by c4bfd01

**Comment by cmeiklejohn** [comment]
This looks good to me, but a core/kv reviewer would be best.

**Comment by slfritchie** [comment]
+1, looks like handy functions to expose to the outside world.

**Pull request merged**